File: backend/app/schemas/cargo.py
"""Versioned cargo data shared by the editor and independent consumers."""
from __future__ import annotations

from typing import Annotated, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

Measure = Annotated[float, Field(ge=0, le=1e12, allow_inf_nan=False)]
Positive = Annotated[float, Field(gt=0, le=1e12, allow_inf_nan=False)]


class CargoModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Dimensions(CargoModel):
    length: Positive | None = None
    width: Positive | None = None
    height: Positive | None = None


class PackagingProperties(CargoModel):
    name: str = Field(min_length=1, max_length=160)
    category: str = Field(default="box", min_length=1, max_length=32)
    dimensions_mm: Dimensions | None = None
    inner_dimensions_mm: Dimensions | None = None
    tare_kg: Measure | None = None
    max_payload_kg: Positive | None = None
    max_gross_kg: Positive | None = None
    max_stack_load_kg: Measure | None = None
    stackable: bool | None = None
    keep_upright: bool | None = None
    can_rotate: bool | None = None
    reusable: bool = False

    @field_validator("name", "category")
    @classmethod
    def nonblank(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("A name is required")
        return value.strip()


class PackagingTemplateIn(PackagingProperties):
    language_labels: dict[str, str] = Field(default_factory=dict)
    active: bool = True
    version: int | None = Field(default=None, ge=1)


class CargoUnit(PackagingProperties):
    id: str = Field(default_factory=lambda: str(uuid4()), min_length=1, max_length=64)
    code: str = Field(min_length=1, max_length=64)
    kind: Literal["package", "ctu"] = "package"
    source: Literal["own", "external", "unknown"] = "unknown"
    parent_id: str | None = Field(default=None, max_length=64)
    template_id: str | None = Field(default=None, max_length=64)
    group_id: str | None = Field(default=None, max_length=64)
    equipment_id: int | None = Field(default=None, gt=0)
    external_reference: str | None = Field(default=None, max_length=120)
    loaded_dimensions_mm: Dimensions | None = None
    measured_gross_kg: Measure | None = None
    # A migrated container line already contributes its tare to legacy totals.
    legacy_goods_id: int | None = None

    @field_validator("id", "parent_id")
    @classmethod
    def valid_identity(cls, value: str | None) -> str | None:
        if value is not None:
            return str(UUID(value))
        return value


class CargoAllocation(CargoModel):
    id: str = Field(default_factory=lambda: str(uuid4()), min_length=1, max_length=64)
    goods_id: int
    unit_id: str = Field(min_length=1, max_length=64)
    quantity: Positive

    @field_validator("id", "unit_id")
    @classmethod
    def valid_identity(cls, value: str) -> str:
        return str(UUID(value))


class CargoManifest(CargoModel):
    schema_version: Literal[1] = 1
    shipment_id: str = Field(default_factory=lambda: str(uuid4()), max_length=64)
    revision: int = Field(default=0, ge=0)
    units: list[CargoUnit] = Field(default_factory=list, max_length=10000)
    allocations: list[CargoAllocation] = Field(default_factory=list, max_length=50000)

    @field_validator("shipment_id")
    @classmethod
    def valid_shipment(cls, value: str) -> str:
        return str(UUID(value))

    @model_validator(mode="after")
    def unique_ids(self):
        for values in ([unit.id for unit in self.units], [unit.code for unit in self.units],
                       [allocation.id for allocation in self.allocations]):
            if len(values) != len(set(values)):
                raise ValueError("Cargo identifiers must be unique")
        return self


class CargoAssessmentIn(CargoModel):
    cargo: CargoManifest
    lines: list[dict] = Field(default_factory=list, max_length=10000)

File: frontend/src/api/cargo.ts
import { apiRequest } from "./client";

export interface CargoDimensions { length: number | null; width: number | null; height: number | null }
export interface PackagingTemplate {
  id: string; name: string; language_labels?: Record<string, string>; category: string;
  dimensions_mm?: CargoDimensions | null; inner_dimensions_mm?: CargoDimensions | null;
  tare_kg?: number | null; max_payload_kg?: number | null; max_gross_kg?: number | null;
  max_stack_load_kg?: number | null; stackable?: boolean | null; keep_upright?: boolean | null;
  can_rotate?: boolean | null; reusable?: boolean; active?: boolean; version?: number;
  builtin?: boolean;
}
export interface CargoUnit extends Omit<PackagingTemplate, "active" | "version" | "language_labels" | "builtin"> {
  code: string; kind: "package" | "ctu"; parent_id: string | null;
  template_id?: string | null; group_id?: string | null; equipment_id?: number | null;
  external_reference?: string; loaded_dimensions_mm?: CargoDimensions | null;
  measured_gross_kg?: number | null; source?: "own" | "external" | "unknown";
  legacy_goods_id?: number | null;
}
export interface CargoAllocation { id: string; goods_id: number; unit_id: string; quantity: number }
export interface CargoManifest {
  schema_version: 1; revision: number; shipment_id?: string;
  units: CargoUnit[]; allocations: CargoAllocation[];
}
export interface CargoGood {
  id: number; name: string; quantity: number | null; unit: string;
  weight_kg: number | null; dimensions_mm?: CargoDimensions | null;
}

export interface CargoAssessment {
  cargo: CargoManifest;
  units: { id: string; calculated_gross_kg: number | null; measured_gross_kg: number | null; complete: boolean; [key: string]: unknown }[];
  totals: { goods_kg: number | null; packaging_kg: number | null; cargo_gross_kg: number | null; transport_tare_kg: number | null; transport_gross_kg: number | null; occupied_volume_m3: number | null; complete: boolean };
  loose: { goods_id: number; quantity: number | null; weight_kg: number | null }[];
  issues: { code: string; unit_id?: string; goods_id?: number; [key: string]: unknown }[];
}
const request = <T,>(path: string, options: RequestInit = {}) => apiRequest<T>(`/cargo${path}`, options);

/** Catalogues and physical units deliberately have different wire contracts. */
function templatePayload(value: Partial<PackagingTemplate>) {
  const keys: (keyof PackagingTemplate)[] = ["name", "category", "language_labels", "dimensions_mm", "inner_dimensions_mm", "tare_kg", "max_payload_kg", "max_gross_kg", "max_stack_load_kg", "stackable", "keep_upright", "can_rotate", "reusable", "active", "version"];
  return Object.fromEntries(keys.filter(key => value[key] !== undefined).map(key => [key, value[key]]));
}

export const cargoApi = {
  templates: () => request<PackagingTemplate[]>("/templates"),
  reusableUnits: (query = "", equipmentId?: number) => request<CargoUnit[]>(`/units?q=${encodeURIComponent(query)}${equipmentId == null ? "" : `&equipment_id=${equipmentId}`}`),
  createTemplate: (value: Omit<PackagingTemplate, "id">) => request<PackagingTemplate>("/templates", { method: "POST", body: JSON.stringify(templatePayload(value)) }),
  updateTemplate: (value: PackagingTemplate) => request<PackagingTemplate>(`/templates/${encodeURIComponent(value.id)}`, { method: "PUT", body: JSON.stringify(templatePayload(value)) }),
  archiveTemplate: (value: PackagingTemplate) => request<PackagingTemplate>(`/templates/${encodeURIComponent(value.id)}?version=${value.version ?? 1}`, { method: "DELETE" }),
  assess: (cargo: CargoManifest, lines: unknown[], signal?: AbortSignal) => request<CargoAssessment>("/v1/assess", { method: "POST", body: JSON.stringify({ cargo, lines }), signal }),
};
